packages/fetchai/connections/p2p_libp2p/aea/dhtpeer/dhtpeer.go

```diff
@@ -0,0 +1,818 @@
+	}
+	err = writeBytesConn(conn, []byte("DONE"))
+	ignore(err)
+
+	addr := string(buf)
+	log.Println("DEBUG connection from ", conn.RemoteAddr().String(),
+		"established for Address", addr)
+
+	// Add connection to map
+	dhtPeer.tcpAddresses[addr] = conn
+	if dhtPeer.addressAnnounced {
+		log.Println("DEBUG announcing tcp client address", addr, "...")
+		err = dhtPeer.registerAgentAddress(addr)
+		if err != nil {
+			log.Println("ERROR while announcing tcp client address", addr, "to the dht:", err)
+			return
+		}
+	}
+
+	for {
+		// read envelopes
+		envel, err := readEnvelopeConn(conn)
+		if err != nil {
+			if err == io.EOF {
+				log.Println("INFO connection closed by client:", err)
+				log.Println("      stoppig...")
+			} else {
+				log.Println("ERROR while reading envelope from client connection:", err)
+				log.Println("      aborting..")
+			}
+			break
+		}
+
+		// route envelope
+		go func() {
+			err := dhtPeer.RouteEnvelope(*envel)
+			ignore(err)
+		}()
+	}
+
+}
+
+// ProcessEnvelope register callback function
+func (dhtPeer *DHTPeer) ProcessEnvelope(fn func(aea.Envelope) error) {
+	dhtPeer.processEnvelope = fn
+}
+
+// RouteEnvelope to its destination
+func (dhtPeer *DHTPeer) RouteEnvelope(envel aea.Envelope) error {
+	target := envel.To
+
+	if target == dhtPeer.myAgentAddress {
+		log.Println("DEBUG route envelope destinated to my local agent ...")
+		for dhtPeer.myAgentReady != nil && !dhtPeer.myAgentReady() {
+			log.Println("DEBUG route agent not ready yet, sleeping for some time ...")
+			time.Sleep(time.Duration(100) * time.Millisecond)
+		}
+		if dhtPeer.processEnvelope != nil {
+			err := dhtPeer.processEnvelope(envel)
+			if err != nil {
+				return err
+			}
+		} else {
+			log.Println("WARN route ProcessEnvelope not set, ignoring envelope", envel)
+		}
+	} else if conn, exists := dhtPeer.tcpAddresses[target]; exists {
+		log.Println("DEBUG route - destination", target, " is a delegate client",
+			conn.RemoteAddr().String())
+		return writeEnvelopeConn(conn, envel)
+	} else {
+		var peerID peer.ID
+		var err error
+		if sPeerID, exists := dhtPeer.dhtAddresses[target]; exists {
+			log.Println("DEBUG route - destination", target, "is a relay client")
+			peerID, err = peer.Decode(sPeerID)
+			if err != nil {
+				log.Println("CRITICAL couldn't parse peer id from relay client id")
+				return err
+			}
+		} else {
+			log.Println("DEBUG route - did NOT found destination address locally, looking for it in the DHT...")
+			peerID, err = dhtPeer.lookupAddressDHT(target)
+			if err != nil {
+				log.Println("ERROR route - while looking up address on the DHT:", err)
+				return err
+			}
+		}
+
+		log.Println("DEBUG route - got peer id for agent address", target, ":", peerID.Pretty())
+
+		log.Println("DEBUG route - opening stream to target ", peerID)
+		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
+		stream, err := dhtPeer.routedHost.NewStream(ctx, peerID, "/aea/0.1.0")
```

Comment:
can we declare this constant (`aea/0.1.0`) on the top please?

Reply:
sure, but people should be careful changing these